
---
# Resolving taxon names from BIEN, GBIF and NCBI with WCSP
---

## Script information  

**BIEN/GBIF data set reformat script (_R_)**  

    BIEN_taxonomy/common_format_creator_vectorized.R  

**NCBI data set reformat scripts (_Bash & Python_)**  

    NCBI_taxonomy/scripts/ncbi_name_extract_V3.py  
    NCBI_taxonomy/scripts/remove_duplicate.py  
    NCBI_taxonomy/scripts/Spermatophyta_clean3.14snakeV3.py  
    NCBI_taxonomy/scripts/Spermatophyta_plnDB_cleanerV1.1.sh  
    NCBI_taxonomy/scripts/Spermatophyta_sp_authority_format_V5.py  
    
**Taxon merger (_R_)**      

    taxonomic_matcherV1.2_MT.R  
    
   _Updated with APG family:_    
   
    taxonomic_matcher.v.1.4.R   
    
  
## Documentation  

For documentation on the scripts creating the common format, please refer to the corresponding markdown file:  

  + *[BIEN/GBIF processor](/BIEN_taxonomy/BIEN_common_format_documentation.md)*
  
  + *[NCBI processor](/NCBI_taxonomy/README.md)*

For documentation on the actual taxon merging procedure, see section [Taxon matching logic](#taxon-matching-logic).


<!--
**Taxonomic matching**  

*taxonomic_matcherV1.2_MT.R* performs the matching of the selected common format resulting from BIEN or NCBI data with the [World Checklist of Selected Plant Families (WCSP)](https://wcsp.science.kew.org/home.do).


##   Workflow for taxon matching logic
![workflow for matching](workflow_matching.png) -->



## Workflow  
<!-- 
A text that serves as general documentation of the matching prodecure that can be used e.g. in a supplement.
-->

### Summary
In order to combine species occurrence and molecular data from databases that follow different taxonomy guidelines, we developed a taxonomy matching procedure that resolves taxon names according to their accepted name in the World Checklist for Selected Plant Families (WCSP)<sup>[1]</sup>, which serves as our reference taxonomy. We standardize taxon names from three databases:  

+ The Botanical Ecology and Information Network ([BIEN](https://biendata.org/))<sup>[2]</sup>  
	
+ The molecular database by the National Center for Biotechnology Information ([NCBI](https://www.ncbi.nlm.nih.gov/))<sup>[3]</sup>  

+ Global Biodiversity Information Facility ([GBIF](https://www.gbif.org/))<sup>[4]</sup> 

Taxon names in BIEN database follow the taxonomy provided by the Taxonomic Name Resolution Service (TNRS)<sup>[5]</sup>, NCBI taxon names follow APG IV<sup>[6]</sup> (also see [APWeb](http://www.mobot.org/MOBOT/research/APweb/)). 

Complete scientific taxon names from BIEN, GBIF or NCBI were split into their components and filled into a common format data frame. We extracted the following fields from taxon names: family, author, generic name, genus hybrid, species epiphet, species hybrid, taxon rank, infraspecific name. In case of missing information, NAs were introduced. Genus or species hybrids were marked with an "x" in the respective column.

<p align="center">
<img src="taxonomy_matching_overview.png"/>  
</p>  


### BIEN/GBIF processor
Required data from the BIEN database include columns *scrubbed_taxon_name_no_author*, *scrubbed_family* and *scrubbed_scrubbed_author*. The processor script splits taxon names into their components, using space as a separator. According to the number of strings that were retrieved from a single taxon name, a certain set of regular expressions-conditions where applied the name, assigning each part of the name to its corresponding column in the common format data  frame. The set of conditions differs according to the number of strings retrieved from a taxon name. Conditions used where the position of an "x" (hybrid marker), the position of string starting with a capital letter, and combinations of both conditions. For a list of all conditions with all possible taxon name cases, see the *[BIEN_common_format_documentation](BIEN_common_format_documentation.md)* and [R script](common_format_creator_vectorized.R). 

GBIF processing includes an extra step that separates author names from scientific taxon names.


### NCBI processor  
Briefly, we used `ncbi_name_extract_V3.py` to extract NCBI taxonomy from NCBI database (generated by [phlawd db maker](https://github.com/blackrim/phlawd_db_maker)) based on insterested columns (e.g., "ncbi_id", "parent_ncbi_id", "scientific_name"); `remove_duplicate.py` is emplyed to remove those duplicated entries, and the taxonomy is further cleaned by `Spermatophyta_plnDB_cleanerV1.1.sh`, then pass to `Spermatophyta_sp_authority_format_v4.py`, which mainly splits taxon authority into a seperated column, and correct the taxon status; and lastly  `Spermatophyta_clean3.14snakeV3.py` is used to proccess all those name strings based on taxonic information categories (e.g., order, family, genus_hybrid,genus, species_hybrid, species, infraspecific_rank, infraspecies, taxon_authority, taxon_rank, ncbi_id) match with format of the taxonomy database from [World Checklist of Selected Plant Families (WCSP)](https://wcsp.science.kew.org/home.do).  

_For more detailed NCBI common format documentation clik [here](/NCBI_taxonomy/README.md)_  

### Taxon matching logic
To prepare the WCSP data for matching, we combined the columns *infraspecific_rank* and *taxon_rank*, then shrank the data set to the same columns present in the input data, set plus a column with the accepted plant name ID.  

Matching is carried out following 3 major steps:

In the first step, a `left_join` (keeping all rows from the input dataframe) of the input data set with the relevant WCSP dataframe is performed, using all columns for matching except for author names. Author names are prone to spelling or punctuation errors, which is why they were excluded here. Taxon names with one matching `accepted_plant_name_ID` were stored as resolved. Cases with more than one match were further processed by performing another left join, this time including author names. Cases with one match were added as resolved, cases with multiple matches were conservatively considered as unmatchable.

In the second step and third step, cases with no matches in step 1 were processed:

2. For taxon names with ranks below the species level and identical species and infraspecific names, we performed a `left_join`, excluding `taxonomic_rank` and `infraspecific` name from matching.  
		
3. For the remaining cases with no initial match, we matched both data frames using all columns except for family.  
		
As before, single and multiple matches were added to the respective data frames in both steps.  
	
The output is a table including all initially entered taxon names, the type of match (or "NA""), and the corresponding `accepted_plant_name_ID` (or "NA"").  

See figure XX for a graphical presentation of the procedure.

<p align="center">
<img src="workflow_matching.png"/>  
</p>  

## References  

[1] WCSP (2020). 'World Checklist of Selected Plant Families. Facilitated by the Royal Botanic Gardens, Kew. Published on the Internet; http://wcsp.science.kew.org/ Retrieved 15 December 2019.'  

[2] Botanical Information and Ecology Network (https://bien.nceas.ucsb.edu/bien/)  

[3] NCBI website: https://www.ncbi.nlm.nih.gov/  

[4] GBIF.org, “GBIF Home Page”, 2020.

[5] Boyle, B. et.al. The taxonomic name resolution service: an online tool for automated standardization of plant names. BMC Bioinformatics. 2013, 14:16. doi:10.1186/1471-2105-14-16  

[6] APG IV. An update of the Angiosperm Phylogeny Group classification for the orders and families of flowering plants: APG IV. Botanical Journal of the Linnean Society, 2016, 181, 1–20. [Link](https://academic.oup.com/botlinnean/article/181/1/1/2416499)  

